SIPS/sip-12.md

```diff
@@ -0,0 +1,118 @@
+
+## Specification
+
+> Formal specifications are written in Typescript. Usage of `CAIP-N` specifications, where `N` is a number, are references to [Chain Agnostic Improvement Proposals](https://github.com/ChainAgnostic/CAIPs).
+
+### Language
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and
+"OPTIONAL" written in uppercase in this document are to be interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt)
+
+### Definitions
+
+> This section is non-normative, and merely recapitulates some definitions from [CAIP-2](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-2.md).
+
+- `ChainId` - a [CAIP-2](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-2.md) string.
+  It identifies a specific chain among all blockchains recognized by the CAIP standards.
+  - `ChainId` consists of a `Namespace` and a `Reference`
+    - `Namespace` - a class of similar blockchains. For example EVM-based blockchains.
+    - `Reference` - a way to identify a concrete chain inside a `Namespace`. For example Ethereum Mainnet or one of its test networks.
+
+- `AccountAddress` - The account address portion of a [CAIP-10](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-10.md) ID.
+
+- `Petname system` - A naming system that finds balance between global, collision-free and memorable names. Please see this [article](http://www.skyhunter.com/marcs/petnames/IntroPetNames.html) on this topic.
+
+### Snap Manifest
+
+This SIP specifies a permission named `endowment:name-lookup`.
+The permission grants a snap the ability to expose an `onNameLookup` export that receives an object with `chainId` and `domain` OR `address` fields.
+
+This permission is specified as follows in `snap.manifest.json` files:
+
+```json
+{
+  "initialPermissions": {
+    "endowment:name-lookup": {
+        "chains": ["eip155:1", "bip122:000000000019d6689c085ae165831e93"],
+    }
+  }
+}
+```
+
+`chains` - An array of CAIP-2 chain IDs that the snap supports. This field is useful for a client in order to avoid unnecessary overhead.
+
+### Snap Implementation
+
+Please see below for an example implementation of the API:
+
+```typescript
+import { OnNameLookupHandler } from "@metamask/snap-types";
+
+export const onNameLookup: OnNameLookupHandler = async ({
+  chainId,
+  domain,
+  address
+}) => {
+  if (domain) {
+    return { resolution: /* Get domain resolution */ }
+  } 
+  
+  if (address) {
+    return { resolution: /* Get address resolution */ };
+  }
+
+  return null;
+};
+```
+
+The type for an `onNameLookup` handler function's arguments is:
+
+```typescript
+type OnNameLookupBaseArgs = { 
+  chainId: ChainId
+}
+
+type DomainLookupArgs = OnNameLookUpBaseArgs & { domain: string, address: undefined };
+type AddressLookupArgs = OnNameLookUpBaseArgs & { address: string, domain: undefined };
```

Comment:
Is TypeScript able to infer the type from this now?

Reply:
Yes